### PR TITLE
Move some cudagraphs checks into C++

### PR DESCRIPTION
**Pull request by mlazos**
Based off of https://github.com/pytorch/pytorch/pull/111094
This + cpp guards improves TIMM geomean optimizer performance by about 20%

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by mlazos** [comment]
@eellison not sure how to fix that cpp linter error; do I need to declare the method somewhere for mypy to see it?

File: torch/_inductor/cudagraph_trees.py [hunk 1]
```diff
@@ -1513,9 +1532,12 @@ def check_invariants(self, inputs: List[Tensor]) -> bool:
         """
 
         # previously managed data pointers remain stable
-        for idx in self.cudagraph_managed_idxs:
-            if inputs[idx].data_ptr() != self.static_input_data_ptrs[idx]:
-                return False
+        # this is on the hot path so moved to C++. equivalent to:
+        # return all(t.data_ptr() == data_ptr for (t, data_ptr) in zip(tensors, data_ptrs))
+        if not torch._C._tensors_data_ptrs_at_indices_equal(
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:1537]: You can add this function to `torch/_C/__init__.pyi.in`

File: torch/_inductor/cudagraph_trees.py [hunk 2]
```diff
@@ -936,30 +962,23 @@ def run_first_inputs(self, new_inputs):
         return outputs
 
     def run(self, new_inputs):
-        if config.triton.fast_path_cudagraph_asserts:
-            self.debug_check_invariants_before_invocation()
+        # self.check_static_inputs_are_stable(new_inputs)
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:965]: why is this commented out ?

File: torch/_inductor/cudagraph_trees.py [hunk 3]
```diff
@@ -787,6 +786,16 @@ def __init__(
             set(wrapped_function.static_input_idxs) | set(self.cudagraph_managed_idxs)
         )
 
+        self.non_static_input_idx: List[int] = [
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:789]: Can you use `LevelList` here ? for both the type annotations

File: torch/_inductor/cudagraph_trees.py [hunk 4]
```diff
@@ -924,6 +933,23 @@ def _copy_input(self, idx, dst, src):
         # TODO - one jit kernel across multiple inputs
         dst.copy_(src)
 
+    def check_static_inputs_are_stable(self, new_inputs):
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:936]: This is never being invoked currently. Intentional?

**Review by eellison — COMMENTED** [review]
Would you add a test where we get an error from static_data_ptr changing? You should be able to trigger it from the mark_static api you added with `guard=False`

File: torch/_inductor/cudagraph_trees.py (hunk 4 above)
Comment (mlazos) [reply torch/_inductor/cudagraph_trees.py:936]: Nope, let me uncomment below

**Review by eellison — APPROVED** [review]

**Pull request closed without merging**